consensus/cometbft/service/block_delay.go

```diff
@@ -0,0 +1,154 @@
+// to default.
+//
+// This is needed because the network may stall for a long time and we don't
+// want to rush in new blocks as the network resumes its operation.
+const (
+	MaxDelayBetweenBlocks = 30 * time.Minute
+
+	// TargetBlockTime is the desired block time.
+	//
+	// Note that it CAN'T be lower than the minimal (floor) block time in the
+	// network, which is comprised of the time to a) propose a new block b)
+	// gather 2/3+ prevotes c) gather 2/3+ precommits.
+	TargetBlockTime = 2 * time.Second
+)
+
+type BlockDelay struct {
+	// - genesis time if height is 0 OR
+	// - block time of the last block if height > 0 and time between blocks is
+	// greater than maxDelayBetweenBlocks.
+	InitialTime time.Time
+
+	// - InitChainRequest.InitialHeight OR
+	// - last block height if time between blocks is greater than maxDelayBetweenBlocks.
+	InitialHeight int64
+
+	// PreviousBlockTime is the time of the previous block.
+	PreviousBlockTime time.Time
+}
+
+// CONTRACT: called only once upon genesis during or after InitChain.
+func BlockDelayUponGenesis(genesisTime time.Time, initialHeight int64) *BlockDelay {
+	return &BlockDelay{
+		InitialTime:       genesisTime,
+		InitialHeight:     initialHeight,
+		PreviousBlockTime: genesisTime,
+	}
+}
+
+// BlockDelayFromBytes converts the bytes to a blockDelay.
+//
+// Expected format:
+//
+//	InitialTime (int64) | InitialHeight (int64) | PreviousBlockTime (int64)
+//	(little endian)
+//
+// Panics if it fails to read from the buffer.
+func BlockDelayFromBytes(
+	bz []byte,
+) *BlockDelay {
+	buf := bytes.NewReader(bz)
+	var initialTime, prevBlockTime int64
+	var initialHeight int64
+
+	err := binary.Read(buf, binary.LittleEndian, &initialTime)
+	if err != nil {
+		panic(fmt.Sprintf("failed to read InitialTime: %v", err))
+	}
+	err = binary.Read(buf, binary.LittleEndian, &initialHeight)
+	if err != nil {
+		panic(fmt.Sprintf("failed to read InitialHeight: %v", err))
+	}
+	err = binary.Read(buf, binary.LittleEndian, &prevBlockTime)
+	if err != nil {
+		panic(fmt.Sprintf("failed to read PreviousBlockTime: %v", err))
+	}
+
+	return &BlockDelay{
+		InitialTime:       time.Unix(initialTime, 0),
+		InitialHeight:     initialHeight,
+		PreviousBlockTime: time.Unix(prevBlockTime, 0),
+	}
+}
+
+// Next returns the duration to wait before proposing the next block.
+func (d *BlockDelay) Next(curBlockTime time.Time, curBlockHeight int64) time.Duration {
+	// Until `timeout_commit` is removed from the CometBFT config,
+	// `FinalizeBlockResponse.NextBlockDelay` can't be exactly 0. If it's set to
+	// 0, then `timeout_commit` from the config will be used, which is not what
+	// we want since we're trying to control the block time.
+	const noDelay = 1 * time.Microsecond
+
+	// Reset the initial time and height if the time between blocks is greater
+	// than maxDelayBetweenBlocks.
+	if curBlockTime.Sub(d.PreviousBlockTime) > MaxDelayBetweenBlocks {
```

Comment:
There is not a case where `Next` is *not* called on every height? If there is such a case at some point in time, then we will need to account for `PreviousBlockHeight` as well. Maybe leaving a comment would here to remember.

Reply:
> There is not a case where Next is not called on every height?

No, it's not how the new `NextBlockDelay` feature works. The ABCI application is always expected to provide value.

Reply:
Please feel free to add a comment, but beware of adding a new source of confusion for developers reading the code.

Reply:
Yes you are right, it might add more harm than value. We can add a comment n the function saying it is expected that this is called at every height and leave it at that. No need to talk abut potential new parameters :D